compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -682,4 +715,59 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             | rustc_transmute::Answer::IfAny(_) => Err(NoSolution),
         }
     }
+
+    pub(super) fn can_define_opaque_ty(&mut self, def_id: DefId) -> bool {
+        let Some(def_id) = def_id.as_local() else { return false; };
+        self.infcx.opaque_type_origin(def_id).is_some()
+    }
+
+    pub(super) fn register_opaque_ty(
```

Comment:
again, handle in old solver, FIXME somewhere

Reply:
Added a `InferCtxt::register_hidden_type_in_new_solver` that asserts that it's being used sanely (it's in the right scope, it's only being used in the new trait solver).